src/memray/reporters/assets/temporal_flamegraph.js

```diff
@@ -0,0 +1,299 @@
+  };
+  var layout = {
+    xaxis: {
+      title: {
+        text: "Time",
+      },
+      rangeslider: {
+        visible: true,
+      },
+    },
+    yaxis: {
+      title: {
+        text: "Memory Size",
+      },
+      tickformat: ".4~s",
+      exponentformat: "B",
+      ticksuffix: "B",
+    },
+  };
+
+  Plotly.newPlot("plot", plot_data, layout, config).then((plot) => {
+    console.assert(active_plot === null);
+    active_plot = plot;
+  });
+}
+
+function showLoadingAnimation() {
+  console.log("showLoadingAnimation");
+  document.getElementById("loading").style.display = "block";
+  document.getElementById("overlay").style.display = "block";
+}
+
+function hideLoadingAnimation() {
+  console.log("hideLoadingAnimation");
+  document.getElementById("loading").style.display = "none";
+  document.getElementById("overlay").style.display = "none";
+}
+
+function refreshFlamegraph(event) {
+  console.log("refreshing flame graph!");
+
+  let request_data = getRangeData(event);
+  console.log("range data: " + request_data);
+
+  if (
+    current_dimensions != null &&
+    JSON.stringify(request_data) === JSON.stringify(current_dimensions)
+  ) {
+    return;
+  }
+
+  console.log("showing loading animation");
+  showLoadingAnimation();
+
+  current_dimensions = request_data;
+
+  console.log("finding range of relevant snapshot");
+
+  let idx0 = 0;
+  let idx1 = memory_records.length - 1;
+
+  if (request_data) {
+    let t0 = new Date(request_data.string1).getTime();
+    for (let i = 0; i < memory_records.length; i++) {
+      if (memory_records[i][0] >= t0) {
+        idx0 = i;
+        break;
+      }
+    }
+
+    idx1 = 0;
+    let t1 = new Date(request_data.string2).getTime();
+    for (let i = memory_records.length - 1; i >= 1; i--) {
+      if (memory_records[i - 1][0] < t1) {
+        idx1 = i;
+        break;
+      }
+    }
+  }
+
+  console.log("start index is " + idx0);
+  console.log("end index is " + idx1);
+  console.log("first possible index is 0");
+  console.log("last possible index is " + (memory_records.length - 1));
+
+  console.log("constructing tree");
+  data = packedDataToTree(packed_data, idx0, idx1);
+
+  console.log("drawing chart");
+  getFilteredChart().drawChart(data);
+  console.log("hiding loading animation");
+  hideLoadingAnimation();
+}
+
+function getRangeData(event) {
+  console.log("getRangeData");
+  let request_data = {};
+  if (event.hasOwnProperty("xaxis.range[0]")) {
+    request_data = {
+      string1: event["xaxis.range[0]"],
+      string2: event["xaxis.range[1]"],
+    };
+  } else if (event.hasOwnProperty("xaxis.range")) {
+    request_data = {
+      string1: event["xaxis.range"][0],
+      string2: event["xaxis.range"][1],
+    };
+  } else if (active_plot !== null) {
+    let the_range = active_plot.layout.xaxis.range;
+    request_data = {
+      string1: the_range[0],
+      string2: the_range[1],
+    };
+  } else {
+    return;
+  }
+  return request_data;
+}
```

Comment:
I was trying to simplify this and came out with this:

```
function getRangeData(event) {
  const range = event?.['xaxis.range'] || event?.['xaxis.range[0]'];
  if (!range && !active_plot) return;
  const [string1, string2] = Array.isArray(range) ? range : [range[0], range[1]];
  return { string1, string2 };
}
```

Unclear if is better, though. Is certainly shorter :)

Reply:
Well, this code came verbatim from your prototype - I'm a bit baffled by why plotly gives us the data in such an inconsistent format in the first place.